12.0 sale channel hooks

kevinkhao opened the pull request.

kevinkhao committed:
[ADD] Hooks

kevinkhao committed:
[IMP] Split hook functionality into mixin, improve readme

kevinkhao committed:
[IMP] Some simplification on channel hook mixin

kevinkhao committed:
[IMP] Name nitpicking

kevinkhao committed:
Change name of trigger_hook to trigger_channel_hook, minor error changes

kevinkhao committed:
[IMP] Change some functionality for hooks
Add licensing

kevinkhao committed:
[IMP] Misc. cleanup, add tests

kevinkhao committed:
make travis happy

kevinkhao committed:
[IMP] Interface change

kevinkhao committed:
make travis happy

kevinkhao committed:
make travis happy

kevinkhao committed:
[IMP] Minor UI: onchange for unticking create_invoice_send_pdf bool

kevinkhao committed:
[IMP] Corrections to content generation
Normalize hook content getter signature
Tests

kevinkhao committed:
[IMP] Make hook calls async
Compute year according to current year for invoice test

sebastienbeau commented on sale_channel_hook/models/sale_channel.py:
```diff
@@ -0,0 +1,47 @@
+class SaleChannel(models.Model):
+    _inherit = "sale.channel"
+
+    auth_token = fields.Char("Secret authentication token")  # DISCUSSION: sécurité
```
use server env for storing this key

sebastienbeau replied:
@kevinkhao https://github.com/OCA/server-env/tree/12.0/server_environment

kevinkhao replied:
done

kevinkhao committed:
[IMP] Add method name in webhook call
Change content for delivery notification: notifies each time packages are sent instead of once all packages are sent
Change payload structure for all modules

kevinkhao committed:
[IMP] Cleanup interface

kevinkhao committed:
make travis happy

kevinkhao committed:
Cleanup interface

kevinkhao committed:
[IMP] Add some demo data for hook api keys and endpoints

sebastienbeau commented on sale_channel_hook_invoice/models/account_invoice.py:
```diff
@@ -0,0 +1,28 @@
+            rec.trigger_channel_hook("create_invoice")
+        return result
+
+    def get_hook_content_create_invoice(self):
```
Only send hook for simple sale case, skip the complex case we are here for a simple implementation

sales = self.mapped("sale_line_ids.order_id")
if sales:
     content = {"sale_name": sales[0].name, "invoice": self.number}

....


else:
     return None

kevinkhao replied:
done, 
+ origin determined through invoice lines->SO lines instead of "origin" field
+ presence of an origin SO is tested before triggering the hook

kevinkhao committed:
[IMP] invoice: find out origin SO through invoice line ids -> corresponding sale order lines instead of origin reference

kevinkhao committed:
Add warning in case there are several SOs for one invoice

sebastienbeau commented on sale_channel_hook/models/sale_channel.py:
```diff
@@ -0,0 +1,47 @@
+                    "endpoint to use this channel's hook"
+                )
+            )
+        # check hook is activated
```
not need as the job only exist if the hook is active

kevinkhao replied:
done. 

Check that the hook is active will be made on the object using the hook (delivery, sale, invoice) instead of the sale channel's job to avoid creating useless jobs

sebastienbeau commented on sale_channel_hook/models/sale_channel.py:
```diff
@@ -0,0 +1,47 @@
+        if not getattr(self, "hook_active_" + hook_name):
+            return
+        payload = {content}
+        signature = self._generate_hook_request_signature(payload)
```
@kevinkhao we should discuss about this

kevinkhao replied:
done. Documentation on what it is and how it's used is now in the readme.

sebastienbeau commented:
@kevinkhao be careful I have forced push a rebase

sebastienbeau committed:
[FIX] add setup

sebastienbeau commented:
TODO improve UI 
![image](https://user-images.githubusercontent.com/1164578/93829124-36d84780-fc6d-11ea-88d0-c9af0c601dd5.png)

kevinkhao committed:
[IMP] Add server environment mixin for channel API keys
Improve documentation on using API keys and using server env

kevinkhao committed:
[IMP] Move checking on whether a hook is active from SaleChannel.send_hook_api_request() to the model using the hook (invoice, delivery...) in order to avoid creating useless jobs

kevinkhao committed:
[IMP] Improve UI for sale channel form view

kevinkhao commented:
> TODO improve UI
> ![image](https://user-images.githubusercontent.com/1164578/93829124-36d84780-fc6d-11ea-88d0-c9af0c601dd5.png)

Done. 
![Screenshot from 2020-09-23 16-27-03](https://user-images.githubusercontent.com/19902174/94026429-f2a68d80-fdb9-11ea-833e-fb16fda34d01.png)

sebastienbeau commented:
Can you put the name as a title (like in Sale Order for example) ?

kevinkhao committed:
[IMP] Visual improvements to sale.channel form view

kevinkhao commented:
done 

![Screenshot from 2020-09-24 16-58-13](https://user-images.githubusercontent.com/19902174/94162684-4206bf80-fe87-11ea-9f36-9fbe17a8b43a.png)

kevinkhao committed:
[IMP] Minor add readme to queue_job_chunk

kevinkhao committed:
Make travis happy

sebastienbeau committed:
[REF] refactor code

kevinkhao committed:
[IMP] Minor fixes and improvements to tests and services

kevinkhao committed:
[IMP] Make security extensible on webhooks, remove old signature generation mechanism

kevinkhao commented:
- improve extensibility of hook api key usage
- in tests, use dispatch instead of creating a service
- minor typo fix

kevinkhao committed:
[IMP] Remove useless bool check and comments

sebastienbeau committed:
[REF] keep the api consistent

sebastienbeau commented on sale_channel_hook/models/sale_channel.py:
```diff
@@ -19,6 +16,10 @@ class SaleChannel(models.Model):
     auth_token = fields.Char("Secret authentication token")
     api_endpoint = fields.Char("Hooks API endpoint")
 
+    def _apply_webhook_security(self, headers, payload, url):
```
The idea is not to remove all the auth code but to implement
- a select field on the channel to choose the auth method (signature ou basic_auth)
- implement the 2 method _auth_basic, auth_signature
- in the method _apply_webhook_security (maybe better to name it _add_weebhook_auth) call the right method depending of the method selected

kevinkhao replied:
Done. Following our discussion about testing requests, I just used tests for the header/content/url of the request and use json=[...], it seems inappropriate to use mock_requests here

kevinkhao committed:
[IMP] Add select field on sale channel to choose security for webhooks, add tests

kevinkhao committed:
[IMP] Directly use JSON in the request instead of dict

kevinkhao committed:
[FIX] Make travis happy

kevinkhao committed:
[IMP] Nitpicking on "none" authentication method
